gcc/cmassoc/graph.1
.TH graph 1 "August 2013" "cmassoc-tools-1.9.1" "Motley Tools"

.SH NAME
graph - print a histogram 

.SH SYNOPSIS
.BR graph
.RI [ options ]
.IR file 
.RI [ file ]
.RI [ ... ]
.RI [> file ]

.PP
.BR graph
.RI [ options ]
.RI [< file ]
.RI [> file ]

.SH DESCRIPTION
Print a text-based histogram on stdout.
The histogram consists of x and y values printed vertically as pairs.
Each pair is followed by a string of hash ('#') characters that represent the magnitude of the y value.
In principle, the program can display an unlimited number of data pairs.

.PP
This program merely displays data.
Data can be generated in many ways and manipulated by selection, sorting and grouping prior to presentation.
It is left to the user to generate, manipulate and interpret the data that is displayed.

.SH OPTIONS
Long options are not supported.
All options are single letters.

.TP
-\fBc\fI pairs\fR
The maximum number of data pairs accepted by the program.
The default is \fB5000\fR.
The range is \fB1\fR to \fBINT_MAX\fR as defined in header limits.h.

.TP
-\fBo\fI [+|-]offset\fR
The data offset expressed as a positive or negative integer value.
All y values are are adjusted by this amount before scaling.
A positive offset shifts data values higher so that negative values can be seen.
A negative offset shits data values lower so that data variations can be accentuated using a larger scale factor.
The default offset is \fB0\fR.
The effective range is \fB-INT_MAX\fR to \fBINT_MAX\fR as defined in header limits.h.

.TP
-\fBs\fI scale\fR
The graph scale expressed as a positive integer value.
All y values are divided by this amount.
The larger this value the shorter the horizontal bars.
The default scale is \fB1\fR.
The range is \fB1\fR to \fBINT_MAX\fR as defined in header limits.h.

.TP
-\fBt\fI title\fR
The graph title.
Enclose the title in quotes or apostrophes if it contains spaces.
The default title is "\fBDistribution\fR".

.TP
.RB - x
Exchange x and y values read from the input.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over previous command line options.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over previous command line options.

.SH ARGUMENTS

.TP
.IR file
The name of a data file.
Data files consist of decimal integer pairs, one pair per line, separated by white space.
A hash ('#') or semicolon (';') causes remaining text on the line to be ignored.

.SH EXAMPLES
The following is an example data file, \fBpowers.txt\fR, containing powers of two.
The first column contains the exponent.
The second column contains the value.
Three comment lines start the file and will be ignored.

.PP
   # ===============
   # powers of two;
   # ---------------
   0 1 
   1 2
   2 4
   3 8
   4 16
   5 32
   6 64
   7 128
   8 256
   9 512

.PP
The following example graphs the file shown above.
The x values appear outside the brackets and the y values appear inside the brackets.
The bars represent the y value but some bars have been truncated in this example because long bars do not format well.

.PP
   # graph powers.txt
   Distribution
        0[     1]#
        1[     2]##
        2[     4]####
        3[     8]########
        4[    16]################
        5[    32]################################
        6[    64]###################################################
        7[   128]###################################################
        8[   256]###################################################
        9[   512]###################################################
   total=1023 level=0 scale=1

.PP
The rext example scales the output to improve the display.
Observe the default graph title of "Distribution".
The total is the sum of y values because this is a histogram program and mathematicians like to know the sample size.
The level is 0 so nothing has been subtracted from y values.
The scale is 16 so each hash represents 16 units but the true y value is still displayed.

.PP
   # graph powers.txt -s 16
   Distribution
        0[     1]
        1[     2]
        2[     4]
        3[     8]
        4[    16]#
        5[    32]##
        6[    64]####
        7[   128]########
        8[   256]################
        9[   512]################################
   total=1023 level=0 scale=16

.SH SEE ALSO
.BR delta ( 1 )

.SH CREDITS
 Charles Maier <cmaier@cmassoc.net>



gcc/cmassoc/graph.c
/*====================================================================*
 *
 *   graph.c - print a histogram;
 *
 *   read x and y integer pairs from stdin or from one or more named
 *   files and print a histogram on stdout; each file prints a new 
 *   histogram; input consists of two integers per line seperated by
 *   whitespace; negative values are converted to zero;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <stdio.h>
#include <ctype.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/cmassoc.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/sintspec.c"
#include "../tools/todigit.c"
#include "../tools/efreopen.c"
#include "../tools/error.c"
#include "../tools/error_on_line.c"
#include "../tools/alert.c"
#endif

/*====================================================================*
 *   program constants; 
 *--------------------------------------------------------------------*/

#define GRAPH_TITLE "Distribution"
#define GRAPH_COUNT 10000

#define GRAPH_INVERT (1 << 0)

/*====================================================================*
 *   program variables; 
 *--------------------------------------------------------------------*/

typedef struct graph 

{ 
	char const * title; 
	signed limit; 
	signed count; 
	signed lower; 
	signed upper; 
	signed shift; 
	signed scale; 
} 

GRAPH; 

/*====================================================================*
 *
 *   void display (signed x [], signed y [], signed count, struct graph * graph) 
 *
 *   display datasets x and y as a histogram; level and scale dataset 
 *   y;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void display (signed x [], signed y [], signed count, struct graph * graph) 

{ 
	printf ("%s\n", graph->title); 
	for (graph->count = 0; graph->count < count; graph->count++) 
	{ 
		signed value = * y; 
		printf ("%6d[%6d]", * x++, * y++); 
		if (value < graph->lower) 
		{ 
			graph->lower = value; 
		} 
		if (value > graph->upper) 
		{ 
			graph->upper = value; 
		} 
		value += graph->shift; 
		while (value > (graph->scale >> 1)) 
		{ 
			putc ('#', stdout); 
			value -= graph->scale; 
		} 
		putc ('\n', stdout); 
	} 
	printf ("count=%d ", graph->count); 
	printf ("lower=%d ", graph->lower); 
	printf ("upper=%d ", graph->upper); 
	printf ("shift=%d ", graph->shift); 
	printf ("scale=%d\n", graph->scale); 
	return; 
} 

/*====================================================================*
 *
 *   void arrange (signed x [], signed y [], signed count);
 *
 *   arrange datasets x and y in ascending order by x values using
 *   a bridge player sort; return no value;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void arrange (signed x [], signed y [], signed count) 

{ 
	signed lower; 
	signed upper; 
	signed value; 
	for (upper = 0; upper < count; upper++) 
	{ 
		for (lower = upper; lower > 0; lower--) 
		{ 
			if (x [lower] < x [lower - 1]) 
			{ 
				value = x [lower]; 
				x [lower] = x [lower - 1]; 
				x [lower - 1] = value; 
				value = y [lower]; 
				y [lower] = y [lower - 1]; 
				y [lower - 1] = value; 
			} 
		} 
	} 
	return; 
} 

/*====================================================================*
 *
 *   void collect (char const * filename, unsigned pairs, struct graph * graph, flag_t flags);
 *
 *   read stdin; parse x and y components; arrange and display;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void collect (char const * filename, signed limit, struct graph * graph, flag_t flags) 

{ 
	signed x [limit]; 
	signed y [limit]; 
	signed index = 0; 
	signed value = 0; 
	signed sign = 0; 
	signed line = 1; 
	signed c = getc (stdin); 
	while ((c != EOF) && (index < limit)) 
	{ 
		if ((c == '#') || (c == ';')) 
		{ 
			do 
			{ 
				c = getc (stdin); 
			} 
			while (nobreak (c)); 
		} 
		if (isspace (c)) 
		{ 
			if (c == '\n') 
			{ 
				line++; 
			} 
			c = getc (stdin); 
			continue; 
		} 
		if (c == '-') 
		{ 
			c = getc (stdin); 
			sign = - 1; 
		} 
		else if (c == '+') 
		{ 
			c = getc (stdin); 
			sign = 1; 
		} 
		else 
		{ 
			sign = 1; 
		} 
		if (!isdigit (c)) 
		{ 
			error_on_line (1, EINVAL, filename, line, "x value"); 
		} 
		value = 0; 
		while (isdigit (c)) 
		{ 
			value *= 10; 
			value += c - '0'; 
			c = getc (stdin); 
		} 
		x [index] = sign * value; 
		while (isblank (c)) 
		{ 
			c = getc (stdin); 
		} 
		if (c == '-') 
		{ 
			c = getc (stdin); 
			sign = - 1; 
		} 
		else if (c == '+') 
		{ 
			c = getc (stdin); 
			sign = 1; 
		} 
		else 
		{ 
			sign = 1; 
		} 
		if (!isdigit (c)) 
		{ 
			error_on_line (1, EINVAL, filename, line, "y value"); 
		} 
		value = 0; 
		while (isdigit (c)) 
		{ 
			value *= 10; 
			value += c - '0'; 
			c = getc (stdin); 
		} 
		y [index] = sign * value; 
		while (isblank (c)) 
		{ 
			c = getc (stdin); 
		} 
		if (nobreak (c)) 
		{ 
			error_on_line (1, EINVAL, filename, line, "line end"); 
		} 
		if (x [index] || y [index]) 
		{ 
			index++; 
		} 
	} 
	if (_anyset (flags, GRAPH_INVERT)) 
	{ 
		arrange (y, x, index); 
		display (y, x, index, graph); 
	} 
	else 
	{ 
		arrange (x, y, index); 
		display (x, y, index, graph); 
	} 
	if (index >= limit) 
	{ 
		error (0, 0, "More than %d data items available.", limit); 
	} 
	return; 
} 

/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	static char const * optv [] = 
	{ 
		"c:o:s:t:x", 
		PUTOPTV_S_FUNNEL, 
		"print a histogram", 
		"c n\tmaximum data points", 
		"o n\tshift data [+|-]", 
		"s n\tscale data", 
		"t s\tgraph title", 
		"x\texchange x and y values", 
		(char const *)(0)
	}; 
	struct graph graph = 
	{ 
		GRAPH_TITLE, 
		0, 
		0, 
		0, 
		0, 
		0, 
		1
	}; 
	signed count = GRAPH_COUNT; 
	flag_t flags = 0; 
	signed c; 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'c': 
			count = (signed)(uintspec (optarg, 1, INT_MAX)); 
			break; 
		case 'o': 
			graph.shift = (signed)(sintspec (optarg, graph.shift)); 
			break; 
		case 's': 
			graph.scale = (signed)(uintspec (optarg, 1, INT_MAX)); 
			break; 
		case 't': 
			graph.title = optarg; 
			break; 
		case 'x': 
			_setbits (flags, GRAPH_INVERT); 
			break; 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	if (!argc) 
	{ 
		collect ("stdin", count, & graph, flags); 
	} 
	while ((argc) && (argv)) 
	{ 
		if (efreopen (* argv, "rb", stdin)) 
		{ 
			collect (* argv, count, & graph, flags); 
		} 
		argc--; 
		argv++; 
	} 
	return (0); 
} 

